### PR TITLE
Feature/search

### PR DESCRIPTION
**Pull request by robot-dreams**
Ready to be reviewed and merged

### CODE REVIEW

**Comment by makosblade** [comment]
LGTM

File: src/main/java/com/kryptnostic/v2/search/SearchApi.java [hunk 1]
```diff
@@ -1,30 +1,70 @@
...
-    String VERSION        = "version";
-    String VERSION_PATH   = "/{" + VERSION + "}";
+    String COUNT        = "count";
+    String COUNT_PATH   = "/{" + COUNT + "}";
 
     /**
-     * Search on stored documents.
+     * The returned set of UUIDs are object ids at which inverted index segments
+     * are stored (encrypted by the AES master key for the channel)
+     *
```
Comment (robot-dreams) [inline src/main/java/com/kryptnostic/v2/search/SearchApi.java:23]: Master key seems to be incorrect.  @geekbeast , can you please verify that this is the "AES cryptoservice for the parent object"?

File: src/main/java/com/kryptnostic/v2/search/SearchApi.java [hunk 2]
```diff
@@ -1,30 +1,70 @@
...
+     * (2) Filter out those inverted index segments corresponding to terms that
+     *     did not appear in the original set of search terms (this filtering
+     *     is necessary due to potential hash collisions)
+     * (3) Collect the object ids appear in the remaining inverted index
+     *     segments; loading and decrypting these objets will result in the
+     *     actual documents containing the search terms
      */
     @POST( CONTROLLER )
-    Set<SearchResult> submitTermQuery( @Body Map<byte[], byte[]> query );
+    Set<UUID> search( @Body Set<byte[]> fheEncryptedSearchTerms );
```
Comment (robot-dreams) [inline src/main/java/com/kryptnostic/v2/search/SearchApi.java]: Consider a more semantically meaningful name?

File: src/main/java/com/kryptnostic/v2/search/SearchApi.java [hunk 3]
```diff
@@ -1,30 +1,72 @@
...
-    String CONTROLLER     = "/search";
-
-    String OBJECT_ID      = "objectId";
-    String OBJECT_ID_PATH = "/{" + OBJECT_ID + "}";
-    String VERSION        = "version";
-    String VERSION_PATH   = "/{" + VERSION + "}";
+    String CONTROLLER              = "/search";
+    String OBJECT_ID               = "objectId";
+    String OBJECT_ID_PATH          = "/{" + OBJECT_ID + "}";
+    String OBJECT_ID_PATH_WITH_DOT = "/{" + OBJECT_ID + ":.+}";
```
Comment (UnsungHero97) [inline src/main/java/com/kryptnostic/v2/search/SearchApi.java]: what's with the `.+`? is this some kind of regex for the URL?

File: src/main/java/com/kryptnostic/v2/storage/models/CreateIndexSegmentRequest.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import com.kryptnostic.v2.constants.Names;
+
+public class CreateIndexSegmentRequest {
+    private final byte[]              address;
+    private final CreateObjectRequest createObjectRequest;
+
+    @JsonCreator
+    public CreateIndexSegmentRequest(
+            @JsonProperty( Names.ADDRESS_FIELD ) byte[] address,
+            @JsonProperty( Names.CREATE_OBJECT_REQUEST_FIELD ) CreateObjectRequest createObjectRequest ) {
```
Comment (UnsungHero97) [inline src/main/java/com/kryptnostic/v2/storage/models/CreateIndexSegmentRequest.java:14]: `CreateObjectRequest` requires a type UUID. what kind of type do you envision this to be? will clients have to register an `invertedIndexSegment` type for this to work?

Comment (makosblade) [inline src/main/java/com/kryptnostic/v2/storage/models/CreateIndexSegmentRequest.java:14]: I think this uses one of the reserved types, INDEX_SEGMENT in TypeUUIDs

File: src/main/java/com/kryptnostic/v2/storage/api/ObjectStorageApi.java [hunk 5]
```diff
@@ -58,7 +60,33 @@
     @POST( CONTROLLER )
     VersionedObjectKey createObject( @Body CreateObjectRequest request );
 
-    @GET( CONTROLLER + LATEST_OBJECT_ID_PATH )
```
Comment (makosblade) [inline src/main/java/com/kryptnostic/v2/storage/api/ObjectStorageApi.java:61]: We definitely still need this line :p

File: src/main/java/com/kryptnostic/v2/search/SearchApi.java (hunk 3 above)
Comment (makosblade) [inline src/main/java/com/kryptnostic/v2/search/SearchApi.java]: This should be equivalent to OBJECT_ID_PATH: "/{objectId}", same with COUNT_PATH_WITH_DOT and COUNT_PATH below

File: src/main/java/com/kryptnostic/v2/storage/models/CreateIndexSegmentRequest.java (hunk 4 above)
Comment (robot-dreams) [inline src/main/java/com/kryptnostic/v2/storage/models/CreateIndexSegmentRequest.java:14]: Yup, @UnsungHero97 and I discussed (it's 00000000-0000-0000-0000-000000000007 if you were curious)

File: src/main/java/com/kryptnostic/v2/storage/api/ObjectStorageApi.java (hunk 5 above)
Comment (robot-dreams) [inline src/main/java/com/kryptnostic/v2/storage/api/ObjectStorageApi.java:61]: Fixed!  I seem to have reverted LATEST_OBJECT_ID_PATH back to OBJECT_ID_PATH when merging--my bad!

Comment (makosblade) [inline src/main/java/com/kryptnostic/v2/storage/api/ObjectStorageApi.java:61]: No worries, thanks!

**Comment by makosblade** [comment]
Still LGTM

**Pull request merged**